src/Build.UnitTests/BackEnd/BuildManager_Tests.cs

```diff
@@ -3805,5 +3807,101 @@ public void MultiProcReentrantProjectWithCallTargetDoesNotFail()
+                     * simulates the ProjectStarted / ProjectEnded events for the unscheduled request and completes it.
+                     *
+                     * However, when the initial build fails in such a way that there is no result for the requested targets,
+                     * the initial BuildResult cannot satisfy the unscheduled submission. In this scenario the Scheduler still completes
+                     * the request but does not simulate the ProjectStarted / ProjectEnded events. It also leaves logging completion to the
+                     * BuildManager.
+                     */
+                    var request1 = new BuildRequestData(testFiles.ProjectFile, new Dictionary<string, string>(), MSBuildConstants.CurrentToolsVersion, new[] {"Build"}, null);
+                    var request2 = new BuildRequestData(testFiles.ProjectFile, new Dictionary<string, string>(), MSBuildConstants.CurrentToolsVersion, new[] {"Build"}, null);
+
+                    /* During builds, msbuild changes the current directory.
+                     * When this test fails, the build never finishes so the current directory never gets restored.
+                     * Force it to restore via the TestEnvironment, otherwise the true test failure gets masked by the current directory not getting restored.
+                     */
+                    env.SetCurrentDirectory(Environment.CurrentDirectory);
+
+                    manager.BeginBuild(parameters);
+
+                    /* Executing async will make the Scheduler aware that there are two identical submissions under consideration.
+                     * Otherwise it builds both sequentually and does not perform the lookahead optimization.
+                     */
+                    var submission1 = manager.PendBuildRequest(request1);
+                    var submission2 = manager.PendBuildRequest(request2);
+
+                    submission1.ExecuteAsync(null, null);
+                    submission2.ExecuteAsync(null, null);
+
+                    submission1.WaitHandle.WaitOne(TimeSpan.FromSeconds(10));
+                    submission2.WaitHandle.WaitOne(TimeSpan.FromSeconds(10));
+
+                    submission1.IsCompleted.ShouldBeTrue();
+                    submission2.IsCompleted.ShouldBeTrue();
+
+                    submission1.BuildResult.Exception.ShouldBeOfType<InvalidProjectFileException>();
+                    submission2.BuildResult.Exception.ShouldBeOfType<InvalidProjectFileException>();
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                }
+                finally
+                {
+                    // Only cleanup the manager if the test did not fail with the expected exception.
+                    // Otherwise EndBuild would hang forever waiting for all submissions to complete.
```

Comment:
I don't think I understand this. Can you elaborate a bit?

Reply:
What about now?